gateway/api/views/files.py

```diff
@@ -276,48 +288,148 @@ def provider_list(self, request):
+                )
+
+            file_storage = FileStorage(
+                username=username,
+                working_dir=working_dir,
+                function_title=function_title,
+                provider_name=provider_name,
+            )
+            result = file_storage.get_file(file_name=requested_file_name)
+            if result is None:
+                return Response(
+                    {"message": "Requested file was not found."},
+                    status=status.HTTP_404_NOT_FOUND,
+                )
+
+            file_wrapper, file_type, file_size = result
+            response = StreamingHttpResponse(file_wrapper, content_type=file_type)
+            response["Content-Length"] = file_size
+            response[
+                "Content-Disposition"
+            ] = f"attachment; filename={requested_file_name}"
+            return response
+
+    @action(methods=["GET"], detail=False, url_path="provider/download")
+    def provider_download(self, request):
+        """
+        It returns a file from provider path:
+            - provider_name/function_title
+        """
+        tracer = trace.get_tracer("gateway.tracer")
+        ctx = TraceContextTextMapPropagator().extract(carrier=request.headers)
+        with tracer.start_as_current_span(
+            "gateway.files.provider_download", context=ctx
+        ):
+            username = request.user.username
+            requested_file_name = sanitize_file_name(
+                request.query_params.get("file", None)
+            )
+            provider_name = sanitize_name(request.query_params.get("provider", None))
+            function_title = sanitize_name(request.query_params.get("function", None))
+            working_dir = WorkingDir.PROVIDER_STORAGE
+
+            if not all([requested_file_name, function_title, provider_name]):
+                return Response(
+                    {
+                        "message": "File name, Qiskit Function title and Provider name are mandatory"  # pylint: disable=line-too-long
+                    },
+                    status=status.HTTP_400_BAD_REQUEST,
+                )
+
+            if not FileStorage.is_valid_extension(requested_file_name):
+                extensions = ", ".join(SUPPORTED_FILE_EXTENSIONS)
+                return Response(
+                    {
+                        "message": f"File name needs to have a valid extension: {extensions}"
+                    },
+                    status=status.HTTP_400_BAD_REQUEST,
+                )
+
+            if not self.user_has_provider_access(request.user, provider_name):
+                return Response(
+                    {"message": f"Provider {provider_name} doesn't exist."},
+                    status=status.HTTP_404_NOT_FOUND,
+                )
+
+            function = self.get_function(
+                user=request.user,
+                function_title=function_title,
+                provider_name=provider_name,
+            )
```

Comment:
Since `get_function` check the user has provider access. Do we want to recheck it again?
[We are not doing it](https://github.com/Qiskit/qiskit-serverless/pull/1547#discussion_r1882822838) in the `download` function

Reply:
This is a different use case because this is the end-point for the providers. Here the check is to clarify if the user has access to the provider not to the function:
- user_has_provider_access
- user_has_access_to_provider_function

Reply:
So, returning to your previous comment that I think I misunderstood it. No, in the `download` end-point is not needed to check if the user has access to the provider. It only needs to have access to the function.